core/sail/federation/src/main/java/org/eclipse/rdf4j/sail/federation/optimizers/FederationJoinOptimizer.java

```diff
@@ -356,7 +370,8 @@ else if (dataset != null) {
 				// fallback to using hasStatement()
 				// but hopefully we narrowed it down to results
 				for (RepositoryConnection member : results) {
-					if (member.hasStatement(subj, pred, obj, true, ctx)) {
+					RepositoryBloomFilter bloomFilter = bloomFilters.apply(member.getRepository());
+					if (bloomFilter == null || bloomFilter.mayHaveStatement(member, subj, pred, obj, ctx)) {
```

Comment:
Ie, I would have presumed the refactoring to go to the following:

if (bloomFilter != null && bloomFilter.mayHaveStatement(...))